A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+
+Options to create an admin server are ordered from the most flexible to the least:
+
+1. The control group: using existing API to create the server;
+2. A function that combines admin services from different packages; **<-RECOMMENDED BY THIS PROPOSAL**
+3. A method of the gRPC server class that adds admin services;
+4. A function that creates the gRPC server object bound with admin services;
+5. A function that creates and starts the gRPC server with admin services.
+
+This proposal will recommend option 2 as the admin interface API (for other alternatives, see rationale section below). The usage snippet of admin interface API in each language:
+
+
+### Java
+
+#### Create An Admin Server With Existing gRPC Java API:
+
+```java
+server = ServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .addService(new ChannelzService())
+        .addService(new CsdsService())
+        .build()
+        .start();
+server.awaitTermination();
+```
+
+
+#### Create An Admin Server With gRPC Java Admin Interface API:
+
+```java
+server = ServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .addServices(AdminInterface.getStandardServices())
+        .build()
+        .start();
+server.awaitTermination();
+```
+
+
+### Golang
+
+#### Create An Admin Server With Existing gRPC Golang API:
+
+```golang
+lis, err := net.Listen("tcp", ":50051")
+if err != nil {
+        log.Fatalf("failed to listen: %v", err)
+}
+defer lis.Close()
+grpcServer := grpc.NewServer()
+channelzService.RegisterChannelzServiceToServer(grpcServer)
+csdsService.RegisterCsdsServiceToServer(grpcServer)
+if err := grpcServer.Serve(lis); err != nil {
+        log.Fatalf("failed to serve: %v", err)
+}
+```
+
+
+#### Create An Admin Server With gRPC Golang Admin Interface API:
+
+```golang
+lis, err := net.Listen("tcp", ":50051")
+if err != nil {
+        log.Fatalf("failed to listen: %v", err)
+}
+defer lis.Close()
+grpcServer := grpc.NewServer(...opts)
+adminServices.RegisterAdminServicesToServer(grpcServer)
+if err := grpcServer.Serve(lis); err != nil {
+        log.Fatalf("failed to serve: %v", err)
+}
+
+```
+
+
+### C++
+
+#### Create An Admin Server With Existing gRPC C++ API:
+
+```cpp
+grpc::ServerBuilder builder;
+builder.RegisterService(grpc::ChannelzService());
+builder.RegisterService(grpc::experimental::CsdsService());
+builder.AddListeningPort(":50051", grpc::ServerCredentials(...));
+std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
+```
+
+
+#### Create An Admin Server With gRPC C++ Admin Interface API:
+
+The admin services will use the gRPC C++ Sync API to offload the responsibility of managing polling and thread from users. In future, they might migrate to the gRPC C++ Callback API.
+
+```cpp
+grpc::ServerBuilder builder;
+grpc::experimental::AddAdminServices(&builder);
```

Comment:
Since this is going to be in a gRFC, there's no reason to use the `experimental` namespace here.

Reply:
TIL. Removed.